lib/common/tests/utils/pcmk_hostname_test.c

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2021 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU Lesser General Public License
+ * version 2.1 or later (LGPLv2.1+) WITHOUT ANY WARRANTY.
+ */
+
+#include <crm_internal.h>
+#include "mock_private.h"
+
+#include <stdarg.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <setjmp.h>
+#include <cmocka.h>
+
+#include <sys/utsname.h>
+
+int __wrap_uname(struct utsname *buf) {
+    int retval = mock_type(int);
+
+    if (retval == 0) {
+        strcpy(buf->nodename, mock_type(char *));
+    }
+
+    return retval;
+}
+
+static void
+uname_succeeded_test(void **state) {
+    char *retval;
+
+    will_return(__wrap_uname, 0);
+    will_return(__wrap_uname, "somename");
+
+    retval = pcmk_hostname();
+    assert_non_null(retval);
+    assert_string_equal("somename", retval);
```

Comment:
The double call of will_return() and assert_\*() looks weird. Is that how cmocka repeats a test?

Reply:
will_return basically enqueues a value that the mocked function (hopefully, as long as you didn't screw up the test) will fetch and return with a mock_type call.  That pair is how you inject a value from the mocked function into the function you are unit testing.  And then the assertions test that your function does the right thing when the mocked function gives it the values you previously enqueued.

Reply:
But we're only calling pcmk_hostname() once, so are we only using one of the will_return() values?

Reply:
In uname_succeeded_test, the first will_return enqueues a 0 and the second enqueues a "somename".  pcmk_hostname is called once, which calls __wrap_uname once.  The 0 is pulled out of the queue by the first mock_type call and becomes retval.  The "somename" is pulled out of the queue by the second mock_type call and becomes buf->nodename.

Reply:
Ah, I get it now -- __wrap_uname() has two substituted values. Let's add comments along the lines of

    will_return(__wrap_uname, 0); // uname() return value
    will_return(__wrap_uname, "somename"); // uname() buf->nodename

and/or comment __wrap_uname() to say what will_return() calls it requires.

Reply:
Added comments.  My hope is that this will all become clearer as we add more tests and more of us add tests.